File: my_app/www/payment/payment.js
let currentBill = null;


// --------------------------------------------------
// SEARCH BILL
// --------------------------------------------------

document.getElementById("search_btn").addEventListener("click", function () {

    const billId = document.getElementById("bill_id").value.trim();

    if (!billId) {
        showMessage("Please enter Bill ID", "error");
        return;
    }

    getBill(billId);

});


// --------------------------------------------------
// GET BILL
// --------------------------------------------------

function getBill(billId) {

    showMessage("Loading...", "info");

    frappe.call({

        method: "my_app.api1.payment.get_bill",

        args: {
            bill_id: billId
        },

        callback: function (response) {

            if (response.message) {

                currentBill = response.message;

                displayBill(currentBill);

                showMessage("", "");

            }

        },

        error: function (error) {

            console.error(error);

            showMessage(
                "Bill not found",
                "error"
            );

        }

    });

}


// --------------------------------------------------
// DISPLAY BILL
// --------------------------------------------------

function displayBill(bill) {

    document.getElementById("bill_section")
        .classList.remove("hidden");


    // Customer

    document.getElementById("customer_name")
        .textContent = bill.cname || "-";

    document.getElementById("customer_phone")
        .textContent = bill.customer_phone || "-";


    // Products

    const tbody =
        document.getElementById("products_body");

    tbody.innerHTML = "";


    bill.products.forEach(function (product) {

        const row = document.createElement("tr");

        row.innerHTML = `
            <td>${product.prod_name}</td>
            <td>${product.quantity}</td>
            <td>₹${formatAmount(product.price)}</td>
            <td>₹${formatAmount(product.total)}</td>
        `;

        tbody.appendChild(row);

    });


    // Summary

    document.getElementById("gross_total")
        .textContent =
        formatAmount(bill.gross_total);


    document.getElementById("discount")
        .textContent =
        formatAmount(bill.discount_applied);


    document.getElementById("total_amount")
        .textContent =
        formatAmount(bill.total_amount);


    document.getElementById("balance_amount")
        .textContent =
        formatAmount(bill.balance_amount);


    // Maximum payment = current balance

    document.getElementById("pay_amount")
        .max = bill.balance_amount;

}


// --------------------------------------------------
// PAY BUTTON
// --------------------------------------------------

document.getElementById("pay_btn")
    .addEventListener("click", function () {

        if (!currentBill) {

            showMessage(
                "Please search for a bill first",
                "error"
            );

            return;
        }


        const payVia =
            document.getElementById("pay_via").value;


        const payAmount =
            parseFloat(
                document.getElementById("pay_amount").value
            );


        // Validate payment method

        if (!payVia) {

            showMessage(
                "Please select a payment method",
                "error"
            );

            return;
        }


        // Validate amount

        if (!payAmount || payAmount <= 0) {

            showMessage(
                "Please enter a valid payment amount",
                "error"
            );

            return;
        }


        // Check against balance

        if (payAmount > currentBill.balance_amount) {

            showMessage(
                "Payment amount cannot be greater than balance",
                "error"
            );

            return;
        }


        createPayment(
            currentBill.name,
            payVia,
            payAmount
        );

    });


// --------------------------------------------------
// CREATE PAYMENT
// --------------------------------------------------

function createPayment(
    billId,
    payVia,
    payAmount
) {

    const payButton =
        document.getElementById("pay_btn");

    payButton.disabled = true;

    payButton.textContent = "Processing...";


    frappe.call({

        method: "my_app.api1.payment.create_payment",

        args: {

            bill_id: billId,

            pay_via: payVia,

            pay_amount: payAmount

        },

        callback: function (response) {

            if (response.message) {

                const result =
                    response.message;


                showSuccess(result);

            }

        },

        error: function (error) {

            console.error(error);

            showMessage(
                "Payment failed",
                "error"
            );

            payButton.disabled = false;

            payButton.textContent = "Pay Now";

        }

    });

}


// --------------------------------------------------
// SUCCESS
// --------------------------------------------------

function showSuccess(result) {

    const message =
        document.getElementById("message");


    message.className = "success";


    message.innerHTML = `

        <h2>Payment Successful 🎉</h2>

        <p>
            Payment ID:
            <strong>${result.payment_id}</strong>
        </p>

        <p>
            Bill ID:
            <strong>${result.bill_id}</strong>
        </p>

        <p>
            Paid Amount:
            <strong>₹${formatAmount(result.paid_amount)}</strong>
        </p>

        <p>
            Remaining Balance:
            <strong>₹${formatAmount(result.balance_amount)}</strong>
        </p>

        <p>
            Status:
            <strong>${result.workflow_state}</strong>
        </p>

    `;


    // Update displayed balance

    document.getElementById("balance_amount")
        .textContent =
        formatAmount(result.balance_amount);


    // Disable payment if fully paid

    if (result.balance_amount <= 0) {

        document.getElementById("pay_btn")
            .disabled = true;

        document.getElementById("pay_btn")
            .textContent = "Fully Paid";

    }
    else {

        document.getElementById("pay_btn")
            .disabled = false;

        document.getElementById("pay_btn")
            .textContent = "Pay Now";

    }


    // Update local bill data

    currentBill.balance_amount =
        result.balance_amount;

}


// --------------------------------------------------
// MESSAGE
// --------------------------------------------------

function showMessage(text, type) {

    const message =
        document.getElementById("message");

    message.textContent = text;

    message.className = type;

}


// --------------------------------------------------
// FORMAT MONEY
// --------------------------------------------------

function formatAmount(amount) {

    return Number(amount || 0)
        .toFixed(2);

}

File: my_app/api1/payment.py
import frappe


@frappe.whitelist(allow_guest=True)
def get_bill(bill_id):

    if not bill_id:
        frappe.throw("Bill ID is required")

    if not frappe.db.exists("Billing", bill_id):
        frappe.throw(f"Bill {bill_id} does not exist")

    bill = frappe.get_doc("Billing", bill_id)

    products = []

    for row in bill.products:
        products.append({
            "prod_id": row.prod_id,
            "prod_name": row.prod_name,
            "quantity": row.quantity,
            "price": row.price,
            "total": row.total
        })

    return {
        "name": bill.name,
        "customer_phone": bill.customer_phone or "",
        "cname": bill.cname or "",

        "products": products,

        "gross_total": float(bill.gross_total or 0),
        "discount_coupon": bill.discount_coupon or "",
        "discount_applied": float(bill.discount_applied or 0),
        "total_amount": float(bill.total_amount or 0),
        "balance_amount": float(bill.balance_amount or 0),

        "workflow_state": bill.workflow_state or ""
    }


@frappe.whitelist(allow_guest=True)
def create_payment(bill_id, pay_via, pay_amount):

    if not bill_id:
        frappe.throw("Bill ID is required")

    if not pay_via:
        frappe.throw("Payment method is required")

    try:
        pay_amount = float(pay_amount)
    except (TypeError, ValueError):
        frappe.throw("Invalid payment amount")

    if pay_amount <= 0:
        frappe.throw("Payment amount must be greater than 0")

    # Get Billing
    bill = frappe.get_doc("Billing", bill_id)

    current_balance = float(bill.balance_amount or 0)

    if current_balance <= 0:
        frappe.throw("This bill has already been fully paid")

    if pay_amount > current_balance:
        frappe.throw(
            f"Payment amount cannot be greater than balance amount ₹{current_balance:.2f}"
        )

    # Calculate new balance
    new_balance = current_balance - pay_amount

    if new_balance <= 0:
        new_balance = 0
        workflow_state = "Paid"
    else:
        workflow_state = "Partially Paid"

    # Create Make Payment
    payment = frappe.get_doc({
        "doctype": "Make Payment",
        "bill_id": bill.name,
        "total_amount": bill.total_amount,
        "balance_amount": new_balance,
        "pay_via": pay_via,
        "pay_amount": pay_amount
    })

    payment.insert(ignore_permissions=True)

    # Update Billing directly
    frappe.db.set_value(
        "Billing",
        bill.name,
        {
            "balance_amount": new_balance,
            "workflow_state": workflow_state
        }
    )

    frappe.db.commit()

    return {
        "success": True,
        "payment_id": payment.name,
        "bill_id": bill.name,
        "paid_amount": pay_amount,
        "balance_amount": new_balance,
        "workflow_state": workflow_state
    }